editor/src/clj/editor/editor_extensions/graph.clj

```diff
@@ -0,0 +1,154 @@
+  (validation/ensure (s/or :nothing #{""} :resource ::validation/node-id) node-id-or-path)
+  (when-not (= node-id-or-path "")
+    (let [node-id (node-id-or-path->node-id node-id-or-path project evaluation-context)
+          resource (g/node-value node-id :resource evaluation-context)
+          ext (:ext (properties/property-edit-type outline-property))]
+      (when (seq ext)
+        (validation/ensure (set ext) (resource/type-ext resource)))
+      resource)))
+
+(def ^:private edit-type-id->value-converter
+  {g/Str {:to identity :from (ensuring-converter string?)}
+   g/Bool {:to identity :from (ensuring-converter boolean?)}
+   g/Num {:to identity :from (ensuring-converter number?)}
+   g/Int {:to identity :from (ensuring-converter int?)}
+   types/Vec2 {:to identity :from (ensuring-converter (s/tuple number? number?))}
+   types/Vec3 {:to identity :from (ensuring-converter (s/tuple number? number? number?))}
+   types/Vec4 {:to identity :from (ensuring-converter (s/tuple number? number? number? number?))}
+   'editor.resource.Resource {:to resource/proj-path :from resource-converter}})
+
+(defn- multi-responds? [^MultiFn multi & args]
+  (some? (.getMethod multi (apply (.-dispatchFn multi) args))))
+
+(defn- property->prop-kw [property]
+  (if (string/starts-with? property "__")
+    (keyword property)
+    (keyword (string/replace property "_" "-"))))
+
+(defn- outline-property [node-id property ec]
+  (when (g/node-instance? (:basis ec) outline/OutlineNode node-id)
+    (let [prop-kw (property->prop-kw property)
+          outline-property (-> node-id
+                               (g/node-value :_properties ec)
+                               (get-in [:properties prop-kw]))]
+      (when (and outline-property
+                 (properties/visible? outline-property)
+                 (edit-type-id->value-converter (properties/edit-type-id outline-property)))
+        (cond-> outline-property
+                (not (contains? outline-property :prop-kw))
+                (assoc :prop-kw prop-kw))))))
+
+(defmulti ext-get (fn [node-id property ec]
+                    [(node-id->type-keyword node-id ec) property]))
+
+(defmethod ext-get [:editor.code.resource/CodeEditorResourceNode "text"] [node-id _ ec]
+  (string/join \newline (g/node-value node-id :lines ec)))
+
+(defmethod ext-get [:editor.resource/ResourceNode "path"] [node-id _ ec]
+  (resource/resource->proj-path (g/node-value node-id :resource ec)))
+
+(defn ext-value-getter
+  "Create 0-arg fn that produces node property value
+
+  Returns nil if there is no getter for node-id+property
+
+  Args:
+    node-id               the node id
+    property              string property name
+    evaluation-context    used evaluation context"
+  [node-id property evaluation-context]
+  (if (multi-responds? ext-get node-id property evaluation-context)
+    #(ext-get node-id property evaluation-context)
+    (if-let [outline-property (outline-property node-id property evaluation-context)]
```

Comment:
This could be `when-let`?

Reply:
Yes. It was like that before, but it's nice to tidy up some code.